runtime/src/cost_tracker.rs

```diff
@@ -199,40 +222,64 @@ impl CostTracker {
 
     fn add_transaction_cost(&mut self, tx_cost: &TransactionCost) {
         let cost = tx_cost.sum();
+        self.add_transaction_execution_cost(tx_cost, cost);
+        saturating_add_assign!(self.account_data_size, tx_cost.account_data_size);
+        saturating_add_assign!(self.transaction_count, 1);
+    }
+
+    fn remove_transaction_cost(&mut self, tx_cost: &TransactionCost) {
+        let cost = tx_cost.sum();
+        self.sub_transaction_execution_cost(tx_cost, cost);
+        self.account_data_size = self
+            .account_data_size
+            .saturating_sub(tx_cost.account_data_size);
+        self.transaction_count = self.transaction_count.saturating_sub(1);
+    }
+
+    /// Apply additional actual execution units to cost_tracker
+    fn add_transaction_execution_cost(&mut self, tx_cost: &TransactionCost, adjustment: u64) {
         for account_key in tx_cost.writable_accounts.iter() {
             let account_cost = self
                 .cost_by_writable_accounts
                 .entry(*account_key)
                 .or_insert(0);
-            *account_cost = account_cost.saturating_add(cost);
+            *account_cost = account_cost
+                .checked_add(adjustment)
+                .unwrap_or(self.account_cost_limit);
         }
-        self.block_cost = self.block_cost.saturating_add(cost);
+        self.block_cost = self
+            .block_cost
+            .checked_add(adjustment)
+            .unwrap_or(self.block_cost_limit);
         if tx_cost.is_simple_vote {
-            self.vote_cost = self.vote_cost.saturating_add(cost);
+            self.vote_cost = self
+                .vote_cost
+                .checked_add(adjustment)
+                .unwrap_or(self.vote_cost_limit);
         }
-        self.account_data_size = self
-            .account_data_size
-            .saturating_add(tx_cost.account_data_size);
-        self.transaction_count = self.transaction_count.saturating_add(1);
     }
 
-    fn remove_transaction_cost(&mut self, tx_cost: &TransactionCost) {
-        let cost = tx_cost.sum();
+    /// Substract extra execution units from cost_tracker
+    fn sub_transaction_execution_cost(&mut self, tx_cost: &TransactionCost, adjustment: u64) {
         for account_key in tx_cost.writable_accounts.iter() {
             let account_cost = self
                 .cost_by_writable_accounts
                 .entry(*account_key)
                 .or_insert(0);
-            *account_cost = account_cost.saturating_sub(cost);
+            *account_cost = account_cost
+                .checked_sub(adjustment)
+                .unwrap_or(self.account_cost_limit);
```

Comment:
Should probably remove the account key if the cost goes back to zero because otherwise "cost_tracker_stats" will report too high of a value for "number_of_accounts"

Reply:
> Should probably remove the account key if the cost goes back to zero because otherwise "cost_tracker_stats" will report too high of a value for "number_of_accounts"

Tracked cost would not change after set to its limit. The accounts hit account_limit need to stay there so no new TX that writes to them would be selected for current block.

Reply:
Oh that's not what I meant.. I wasn't referring to underflow. I was referring to when a transaction increments the cost for an account it uses, and then that transaction is removed, the account cost will go to zero but it doesn't get removed from the map. Then, when we check the size of the account cost map, we will incorrectly report a higher number of accounts in the metrics.

Reply:
Gotcha gotcha. Maybe removing zero-cost accounts from Hashmap is an expense to be avoid, same account can be added back into Hashmap by next transaction. Polly it is easier to add a function to count number of non-zero accounts for metrics reporting, as in this [commit](79782d73703d24e0f2d362c27d08a097ba8b3b05).

Reply:
Your fix works too, thanks!